### ISO Specific Information

Mostly related to how version is represented and rolecode details. 

Are the rolecode URLs still available? How many are from an unexpected source? 

Do any responses reference code lists from different sources? 

How many responses (or host systems) contain a reference to a codelist that's no longer available? 

And what informational content will be affected, ie are the spatial reference blocks more likely to reference a dead code list link?


How are ISO versions represented? Is it clear enough to understand how it might be structured?

(So, no. What matters more for parsing and reading the XML is more keyed to the schema definition. SMAP is interesting in that there's a mimetype (unnecessary) but I think not a custom schema - it's a standardized structure for SMAP in ISO but is just about how to use these elements in this way to reduce the variability. It should just be ISO and validated that way; knowing it's SMAP doesn't affect anything *if* a client is built to the spec, not the server (haha. hahaha.).) But no, the version string gives you a bit of a clue to where some of those other, similar uses of ISO. This is different than INSPIRE, right, that one is an additional namespace so it is validated for ISO and for INSPIRE structure.

It is ISO dialects.


In [1]:
%matplotlib inline
import pandas as pd
import json as js  # name conflict with sqla
import sqlalchemy as sqla
from sqlalchemy.orm import sessionmaker
from IPython.display import display

In [2]:
# grab the clean text from the rds
with open('../local/big_rds.conf', 'r') as f:
    conf = js.loads(f.read())

# our connection
engine = sqla.create_engine(conf.get('connection'))

In [ ]:
df = pd.read_sql(sql, engine)
with pd.option_context('display.max_colwidth', 1000):
    display(df[:10])

### Versions

Where to find ISO version information:



| Standard Name                                                       |  Version                                                                | 
|---------------------------------------------------------------------|-------------------------------------------------------------------------| 
| //*/metadataStandardName/CharacterString                            |  //*/metadataStandardVersion/CharacterString                            | 
| //*/seriesMetadata/MD_Metadata/metadataStandardName/CharacterString |  //*/seriesMetadata/MD_Metadata/metadataStandardVersion/CharacterString | 
| //*/seriesMetadata/MI_Metadata/metadataStandardName/CharacterString |  //*/seriesMetadata/MI_Metadata/metadataStandardVersion/CharacterString | 


This is just the version info for single metadata records (not the DS proper, but the MI/MD contained by a DS). [Speculation] This bit of information is not often included in the CSW or OAI-PMH getrecords outputs (those never track as valid ISO and certainly not at all complete ISO compared to what can be accessed from other parts of those larger portal systems). That last is important? or not, depends on whether we think CSW or OAI-PMH is functional (slow, so painfully slow in current implementations). They do not track as first-class dev efforts - must have it but not key to the system so not well-built and not well-maintained. But, back one of the points, did not validate any of those internal ISO records (if we have any and we do) and, as noted previously, the specs for XML validation allow for nested vaildation but that is not implemented. PParse, part of the Apache Xerces package, is (been a while since I had to go through that eval) one of the view available free options and it does not support that kind of validation.

In [7]:
# the set of unique version strings

sql = """
select distinct standard_name, standard_version
from iso_versions
order by standard_name, standard_version;
"""

df = pd.read_sql(sql, engine)
with pd.option_context('display.max_rows', 75):
    display(df[:61])

standard_name  \
0   ANZLIC Metadata Profile: An Australian/New Zea...   
1                           DM - Regole tecniche RNDT   
2                                              Gemini   
3                                              GEMINI   
4             INSPIRE Implementing Rules for Metadata   
5                                            ISO19115   
6                                            ISO19115   
7                                           ISO 19115   
8                                           ISO 19115   
9                                           ISO 19115   
10                                    ISO 19115/19139   
11                                        ISO 19115-2   
12                               ISO 19115:2003/19139   
13                          ISO 19115/2003/Cor.1:2006   
14                            ISO 19115:2003 (GDI-BW)   
15  ISO 19115:2003 - Geographic information - Meta...   
16  ISO 19115-2 Geographic information - Metadata ...   
17  ISO 19115-2 Geographic information -- Metadata...   
18  ISO 19115-2 Geographic information Metadata Pa...   
19  ISO 19115-2 Geographic information - Metadata ...   
20  ISO 19115-2 Geographic information - Metadata ...   
21  ISO 19115-2 Geographic Information - Metadata ...   
22  ISO 19115-2 Geographic Information - Metadata ...   
23  ISO 19115-2 Geographic Information - Metadata ...   
24  ISO 19115-2 Geographic Information - Metadata ...   
25  ISO 19115-2 Geographic Information - Metadata ...   
26  ISO 19115-2 Geographic information - Metadata ...   
27  ISO 19115-2 Geographic Information - Metadata ...   
28  ISO 19115-2 Geographic Information - North Ame...   
29        ISO 19115 Geographic information - Metadata   
30        ISO 19115 Geographic Information - Metadata   
31        ISO 19115 Geographic Information - Metadata   
32  ISO 19115 Geographic information - Metadata - ...   
33  ISO 19115 Geographic information Metadata; WIS...   
34                       ISO 19115/SeaDataNet profile   
35                              ISO 19115 (UK GEMINI)   
36                                           ISO19119   
37                                          ISO 19119   
38              ISO 19139/19115 Metadata for Datasets   
39          ISO 19139/19119 Metadata for Web Services   
40  ISO 19139 Geographic Information - Metadata - ...   
41                                      ISO-NAP-USGIN   
42                                      ISO-NAP-USGIN   
43                                          ISO-USGIN   
44                  MEDIN Discovery metadata standard   
45                  MEDIN Discovery metadata standard   
46                  MEDIN Discovery metadata standard   
47                  MEDIN Discovery Metadata Standard   
48                  MEDIN Discovery Metadata Standard   
49                  MEDIN Discovery Metadata Standard   
50                  MEDIN Discovery Metadata Standard   
51                                     NAP - Metadata   
52                     NERC profile of ISO:19115:2003   
53                      NERC profile of ISO19115:2003   
54                      NERC profile of ISO19115:2003   
55  North American Profile of ISO 19115:2003 - Geo...   
56  North American Profile of ISO19115:2003 - Geog...   
57  North American Profile of ISO 19115 Geographic...   
58                                          UK GEMINI   
59                                          UK GEMINI   
60                          WMO Core Metadata Profile   

                                     standard_version  
0                                                 1.1  
1                                    10 novembre 2011  
2                                                 2.1  
3                                                 2.1  
4                                                 1.2  
5                                                 1.0  
6                                     2003/Cor.1:2006  
7                                                2003  
8

In [11]:
# let's go with pct of all iso per version string, out of 19,689
sql = """
with j as
(
	with i as
	(
		select d.response_id, jsonb_array_elements(d.identity::jsonb) ident
		from identities d
		where d.identity is not null
	), a as (
		select count(*) as total
		from responses r join i on i.response_id = r.id
		where i.ident->>'protocol' = 'ISO'
	)

	select r.id, a.total
	from responses r join i on i.response_id = r.id
		natural inner join a 
	where i.ident->>'protocol' = 'ISO'
), v as 
(
	select response_id,
		standard_name || '; ' || standard_version as the_version
	from iso_versions
)

select v.the_version, count(v.response_id) as num,
	round(count(v.response_id) / max(j.total)::numeric * 100.0, 2) as pct_of_iso
from responses r join v on v.response_id = r.id
	join j on j.id = r.id
group by v.the_version
order by pct_of_iso DESC;
"""

df = pd.read_sql(sql, engine)
with pd.option_context('display.max_rows', 55, 'display.max_colwidth', 500):
    display(df[:54])
    
# the sum is 200 shy 
# TODO: verify that count (not from versions unless there're duplicate entries?)

the_version  \
0                                          ISO 19115-2 Geographic Information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
1                                                                                                       ISO 19115 Geographic Information - Metadata; 2009-02-15   
2                                             ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for imagery and gridded data; ISO 19115-2:2009(E)   
3                                                                                                                                     ISO19115; 2003/Cor.1:2006   
4                                                                                                                                                ISO-USGIN; 1.2   
5                                                                                                                                               ISO 19115; 2003   
6                                                                                                                                          ISO-NAP-USGIN; 1.1.4   
7                                             ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
8                                                                                                              MEDIN Discovery Metadata Standard; Version 2.3.5   
9                                                                                                               ISO 19139/19119 Metadata for Web Services; 2005   
10                                                                                                       ISO 19115 Geographic Information - Metadata; ISO 19115   
11                                                                                                                             ISO 19115 (UK GEMINI); 1.0 (2.2)   
12                                                                                               ISO 19115 Geographic information - Metadata; ISO 19115:2003(E)   
13                                                                                                                           NERC profile of ISO19115:2003; 1.0   
14                                         ISO 19115-2 Geographic information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
15                                  ISO 19115-2 Geographic Information - North American Profile Metadata - Data with Biological Extensions; ISO 19115-2:2009(E)   
16                                                                                                                                    ISO 19115:2003/19139; 1.0   
17                                                  ISO 19115 Geographic information - Metadata - Converted from Data.gov legacy DMS format.; ISO 19115:2003(E)   
18                                                                                                                     MEDIN Discovery metadata standard; 2.3.4   
19                                                                                                                                 ISO 19115:2003 (GDI-BW); 1.2   
20                                                                             ISO 19139 Geographic Information - Metadata - Implementation Specification; 2007   
21                                                                                                                 INSPIRE Implementing Rules for Metadata; 1.2   
22                                            ISO 19115-2 Geographic information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009   
23                                                                                                                                                  GEMINI; 2.1   
24                                                                       

In [ ]:
# pct of the iso responses, ordered by number of responses for the version string,
# per host

sql = """
with j as
(
	with i as
	(
		select d.response_id, jsonb_array_elements(d.identity::jsonb) ident
		from identities d
		where d.identity is not null
	), a as (
		select r.host, count(r.id) as num_per_host
		from responses r join i on i.response_id = r.id
		where i.ident->>'protocol' = 'ISO'
		group by r.host
	)

	select r.id, r.host, a.num_per_host
	from responses r join i on i.response_id = r.id
		join a on a.host = r.host
	where i.ident->>'protocol' = 'ISO'
), v as 
(
	select response_id,
		standard_name || '; ' || standard_version as the_version
	from iso_versions
)

select v.the_version, j.host, --count(j.id) as num_iso_responses,
	round(count(j.id) / max(j.num_per_host)::numeric * 100.0, 2) as pct_of_iso
from responses r join v on v.response_id = r.id
	join j on j.id = r.id
group by v.the_version, j.host
--order by j.host, pct_of_iso DESC;
order by count(v.response_id) DESC;
"""
    
df = pd.read_sql(sql, engine)

In [8]:
with pd.option_context('display.max_rows', 150):
    display(df[:145])

the_version  \
0    ISO 19115-2 Geographic Information - Metadata ...   
1    ISO 19115 Geographic Information - Metadata; 2...   
2                                       ISO-USGIN; 1.2   
3                                      ISO 19115; 2003   
4                                 ISO-NAP-USGIN; 1.1.4   
5    ISO 19115-2 Geographic Information - Metadata ...   
6    ISO 19115-2 Geographic Information - Metadata ...   
7    ISO 19115 Geographic Information - Metadata; 2...   
8    ISO 19115-2 Geographic Information - Metadata ...   
9                            ISO19115; 2003/Cor.1:2006   
10                           ISO19115; 2003/Cor.1:2006   
11     ISO 19139/19119 Metadata for Web Services; 2005   
12    MEDIN Discovery Metadata Standard; Version 2.3.5   
13   ISO 19115-2 Geographic Information - Metadata ...   
14                           ISO19115; 2003/Cor.1:2006   
15   ISO 19115 Geographic Information - Metadata; I...   
16   ISO 19115-2 Geographic Information - Metadata ...   
17   ISO 19115-2 Geographic Information - Metadata ...   
18   ISO 19115-2 Geographic Information - Metadata ...   
19   ISO 19115-2 Geographic Information - Metadata ...   
20                    ISO 19115 (UK GEMINI); 1.0 (2.2)   
21   ISO 19115-2 Geographic Information - Metadata ...   
22   ISO 19115-2 Geographic Information - Metadata ...   
23                  NERC profile of ISO19115:2003; 1.0   
24   ISO 19115-2 Geographic Information - Metadata ...   
25   ISO 19115-2 Geographic Information - Metadata ...   
26                    ISO 19115 (UK GEMINI); 1.0 (2.2)   
27   ISO 19115-2 Geographic Information - Metadata ...   
28   ISO 19115-2 Geographic Information - Metadata ...   
29   ISO 19115-2 Geographic Information - Metadata ...   
30   ISO 19115-2 Geographic Information - Metadata ...   
31   ISO 19115-2 Geographic Information - Metadata ...   
32   ISO 19115 Geographic information - Metadata; I...   
33                           ISO 19115:2003/19139; 1.0   
34   ISO 19115-2 Geographic Information - North Ame...   
35   ISO 19115-2 Geographic Information - Metadata ...   
36   ISO 19115-2 Geographic information - Metadata ...   
37   ISO 19115 Geographic information - Metadata; I...   
38   ISO 19115 Geographic Information - Metadata; I...   
39   ISO 19115 Geographic information - Metadata - ...   
40   ISO 19115-2 Geographic information - Metadata ...   
41                        ISO 19115:2003 (GDI-BW); 1.2   
42                                      ISO-USGIN; 1.2   
43            MEDIN Discovery metadata standard; 2.3.4   
44   ISO 19115-2 Geographic Information - Metadata ...   
45   ISO 19115-2 Geographic Information - Metadata ...   
46                           ISO19115; 2003/Cor.1:2006   
47        INSPIRE Implementing Rules for Metadata; 1.2   
48   ISO 19139 Geographic Information - Metadata - ...   
49    MEDIN Discovery Metadata Standard; Version 2.3.5   
50   ISO 19115 Geographic information - Metadata; I...   
51            MEDIN Discovery metadata standard; 2.3.4   
52                                         GEMINI; 2.1   
53            MEDIN Discovery metadata standard; 2.3.8   
54                                     ISO 19115; 2003   
55   ISO 19115-2 Geographic information - Metadata ...   
56   ISO 19115-2 Geographic Information - Metadata ...   
57   ANZLIC Metadata Profile: An Australian/New Zea...   
58   ISO 19115-2 Geographic Information - Metadata ...   
59   ISO 19115 Geographic information - Metadata - ...   
60              WMO Core Metadata Profile; 1.0 (draft)   
61   ISO 19115-2 Geographic Information - North Ame...   
62                           ISO19115; 2003/Cor.1:2006   
63                     ISO 19115/19139; 2003/cor1:2006   
64   ISO 19115-2 Geographic Information - Metadata ...   
65   ISO 19115-2 Geographic Information - Metadata ...   
66                                      UK GEMINI; 2.2   
67   ISO 19115-2 Geographic Information - Metadata ...   
68                  NERC profile of ISO19

In [12]:
sql = """
with j as
(
	with i as
	(
		select d.response_id, jsonb_array_elements(d.identity::jsonb) ident
		from identities d
		where d.identity is not null
	), a as (
		select r.host, count(r.id) as num_per_host
		from responses r join i on i.response_id = r.id
		where i.ident->>'protocol' = 'ISO'
		group by r.host
	)

	select r.id, r.host, a.num_per_host
	from responses r join i on i.response_id = r.id
		join a on a.host = r.host
	where i.ident->>'protocol' = 'ISO'
), v as 
(
	select response_id,
		standard_name || '; ' || standard_version as the_version
	from iso_versions
)

select v.the_version, j.host, --count(j.id) as num_iso_responses,
	round(count(j.id) / max(j.num_per_host)::numeric * 100.0, 2) as pct_of_iso
from responses r join v on v.response_id = r.id
	join j on j.id = r.id
group by v.the_version, j.host
--order by j.host, pct_of_iso DESC;
order by j.host, pct_of_iso DESC;
"""
    
df = pd.read_sql(sql, engine)
with pd.option_context('display.max_rows', 150, 'display.max_colwidth', 400):
    display(df[:145])

the_version  \
0                                                                                                                               ISO 19115-2; ISO 19115-2:2009(E)   
1                                              ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
2                                           ISO 19115-2 Geographic Information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
3                                                                                                        ISO 19115 Geographic Information - Metadata; 2009-02-15   
4                                                                                                                                      ISO 19115:2003/19139; 1.0   
5                                                                                                 ISO 19115 Geographic information - Metadata; ISO 19115:2003(E)   
6                                                                                                         ISO 19115 Geographic Information - Metadata; ISO 19115   
7                                                                                                                                                 ISO-USGIN; 1.2   
8                                              ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for imagery and gridded data; ISO 19115-2:2009(E)   
9                                                    ISO 19115 Geographic information - Metadata - Converted from Data.gov legacy DMS format.; ISO 19115:2003(E)   
10                                                                                            ISO 19115:2003 - Geographic information - Metadata; ISO 19115:2003   
11                                          ISO 19115-2 Geographic information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
12                                                                                                                   ISO 19139/19115 Metadata for Datasets; 2003   
13                                                                                                                                                 ISO19115; 1.0   
14                                                                                                                                            ISO-NAP-USGIN; 1.1   
15                                          ISO 19115-2 Geographic Information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2013(E)   
16                                             ISO 19115-2 Geographic information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009   
17                                                                              ISO 19139 Geographic Information - Metadata - Implementation Specification; 2007   
18                                          ISO 19115-2 Geographic Information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
19                                             ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for imagery and gridded data; ISO 19115-2:2009(E)   
20                                                                                                                                     ISO19115; 2003/Cor.1:2006   
21                                          ISO 19115-2 Geographic Information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
22                                             ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
23                                                                                                                                     ISO 19115:2003/19139; 1.0   
24                                               

### Codelists


In [13]:
# how many unique codelist references (before the hash only)

sql = """
with a as
(
	select replace(left(codelist, strpos(codelist, '#')), '#', '') as codelist
	from codelists
)
select distinct codelist
from a
where codelist != ''
order by codelist;
"""

df = pd.read_sql(sql, engine)
with pd.option_context('display.max_rows', 60, 'display.max_colwidth', 400):
    display(df[:52])

codelist
0                                               file://Agi-s000001/afdelingen/GSMS/Extern/Applicaties_en_Tools/Databeheer/Metadata/GeoMDE_22/resources/customCodelists_new.xml
1                                                                      file://appconf.intranet.rws.nl/Appconf/metadatatools/Metadatamaker/config_bestanden/customCodelists.xml
2                                                                                      file://Shr-ipvw-gpr001/metadatatools/Metadatamaker/config_bestanden/customCodelists.xml
3                                                                                                          http://adl.brs.gov.au/anrdl/resources/codeList/codeList20120313.xml
4                                                                                                                   http://asdd.ga.gov.au/asdd/profileinfo/GAScopeCodeList.xml
5                                                                                                                      http://asdd.ga.gov.au/asdd/profileinfo/gmxCodelists.xml
6                                                                                     http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/Codelist/gmxCodelists.xml
7                                                                                  http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/Codelist/ML_gmxCodelists.xml
8                                                                               http://aws2.caris.com/sfs/schemas/iso/19139/CARIS_20120814/resources/Codelist/gmxCodelists.xml
9                                                                               http://aws2.caris.com/sfs/schemas/iso/19139/CARIS_20120814/resources/Codelist/srvCodelists.xml
10                                                                                                       http://data.daff.gov.au/anrdl/resources/codeList/codeList20120313.xml
11                                                                                                             http://geobrain.laits.gmu.edu/CatalogXML/resources/codeList.xml
12                                                                                                                      http://mdtranslator.adiwg.org/api/codelists?format=xml
13                                                                                                        http://nap.geogratis.gc.ca/metadata/register/napMetadataRegister.xml
14                                                                                           http://schemas.opengis.net/iso/19139/20070417/resources/Codelist/gmxCodelists.xml
15                                                                                                             http://seadata.bsh.de/isoCodelists/sdnCodelists/csrCodeList.xml
16                                                                                             http://seadatanet.maris2.nl/isocodelists/sdncodelists/edmo-edmerp-Codelists.xml
17                                                                                                      http://standards.iso.org/iso/19139/resources/codelist/gmxCodelists.xml
18                                                              http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resources/codelist/gmxCodelists.xml
19                                                              http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resources/Codelist/gmxCodelists.xml
20                                                           http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resources/codelist/ML_gmxCodelists.xml
21                                                          http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resources/Codelist/ML_gmx Codelists.xml
22                                                           http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resou

In [18]:
# how many of those are http or https links?
# AND some are casing diffs only, so lowercase all
sql = """
with a as
(
	select lower(replace(left(codelist, strpos(codelist, '#')), '#', '')) as codelist
	from codelists
)
select distinct codelist
from a
where codelist ilike 'http%%'
order by codelist;
"""

df = pd.read_sql(sql, engine)
with pd.option_context('display.max_rows', 60, 'display.max_colwidth', 400):
    display(df[:46])

codelist
0                                                                                                          http://adl.brs.gov.au/anrdl/resources/codelist/codelist20120313.xml
1                                                                                                                   http://asdd.ga.gov.au/asdd/profileinfo/gascopecodelist.xml
2                                                                                                                      http://asdd.ga.gov.au/asdd/profileinfo/gmxcodelists.xml
3                                                                                     http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/codelist/gmxcodelists.xml
4                                                                                  http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/codelist/ml_gmxcodelists.xml
5                                                                               http://aws2.caris.com/sfs/schemas/iso/19139/caris_20120814/resources/codelist/gmxcodelists.xml
6                                                                               http://aws2.caris.com/sfs/schemas/iso/19139/caris_20120814/resources/codelist/srvcodelists.xml
7                                                                                                        http://data.daff.gov.au/anrdl/resources/codelist/codelist20120313.xml
8                                                                                                              http://geobrain.laits.gmu.edu/catalogxml/resources/codelist.xml
9                                                                                                                       http://mdtranslator.adiwg.org/api/codelists?format=xml
10                                                                                                        http://nap.geogratis.gc.ca/metadata/register/napmetadataregister.xml
11                                                                                           http://schemas.opengis.net/iso/19139/20070417/resources/codelist/gmxcodelists.xml
12                                                                                                             http://seadata.bsh.de/isocodelists/sdncodelists/csrcodelist.xml
13                                                                                             http://seadatanet.maris2.nl/isocodelists/sdncodelists/edmo-edmerp-codelists.xml
14                                                                                                      http://standards.iso.org/iso/19139/resources/codelist/gmxcodelists.xml
15                                                              http://standards.iso.org/ittf/publiclyavailablestandards/iso_19139_schemas/resources/codelist/gmxcodelists.xml
16                                                          http://standards.iso.org/ittf/publiclyavailablestandards/iso_19139_schemas/resources/codelist/ml_gmx codelists.xml
17                                                           http://standards.iso.org/ittf/publiclyavailablestandards/iso_19139_schemas/resources/codelist/ml_gmxcodelists.xml
18                                                           http://standards.iso.org/ittf/publ.*lyavailablestandards/iso_19139_schemas/resources/codelist/ml_gmxcodelists.xml
19                                                                                                        http://vocab.nerc.ac.uk/isocodelists/sdncodelists/cdicsrcodelist.xml
20                                                                                                         http://vocab.nerc.ac.uk/isocodelists/sdncodelists/edmedcodelist.xml
21                                                                                                          http://vocab.nerc.ac.uk/isocodelists/sdncodelists/gmxcodelists.xml
22     http://webarchive.nationalarchives.gov.uk/20130402145952/http://standards.iso.org/ittf/publiclyavailablestandards/iso_19139_schemas/re

In [20]:
# and a super quick linkrot check
import requests

codes = []
for d in df.itertuples():
    url = d[1]
    try:
        rsp = requests.get(url, timeout=30)
        status_code = rsp.status_code
    except Exception as ex:
        print url, ex
        status_code = 900
    
    codes.append((url, status_code))
    
    
    

http://adl.brs.gov.au/anrdl/resources/codelist/codelist20120313.xml ('Connection aborted.', gaierror(8, 'nodename nor servname provided, or not known'))
http://www.ec.gc.ca/data_donnees/standards/schemas/napec ('Connection aborted.', error(54, 'Connection reset by peer'))
http://www.tc211.org/iso19139/resources/codelist.xml ('Connection aborted.', gaierror(8, 'nodename nor servname provided, or not known'))


In [22]:
from collections import defaultdict
statuses = defaultdict(int)
for url, code in codes:
    print code, url
    statuses[code] += 1

900 http://adl.brs.gov.au/anrdl/resources/codelist/codelist20120313.xml
404 http://asdd.ga.gov.au/asdd/profileinfo/gascopecodelist.xml
404 http://asdd.ga.gov.au/asdd/profileinfo/gmxcodelists.xml
200 http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/codelist/gmxcodelists.xml
200 http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/codelist/ml_gmxcodelists.xml
200 http://aws2.caris.com/sfs/schemas/iso/19139/caris_20120814/resources/codelist/gmxcodelists.xml
200 http://aws2.caris.com/sfs/schemas/iso/19139/caris_20120814/resources/codelist/srvcodelists.xml
404 http://data.daff.gov.au/anrdl/resources/codelist/codelist20120313.xml
404 http://geobrain.laits.gmu.edu/catalogxml/resources/codelist.xml
200 http://mdtranslator.adiwg.org/api/codelists?format=xml
404 http://nap.geogratis.gc.ca/metadata/register/napmetadataregister.xml
404 http://schemas.opengis.net/iso/19139/20070417/resources/codelist/gmxcodelists.xml
404 http://seadata.bsh.de/isocodelists/sdncodelists/csrcod

In [23]:
statuses

defaultdict(<type 'int'>, {200: 14, 404: 25, 900: 3})

In [24]:
len(codes)

42

In [28]:
for k, v in statuses.iteritems():
    print k, '{:.2f}%'.format(v / float(len(codes)) * 100.)

200 33.33%
404 59.52%
900 7.14%


####So how many responses contain an inaccessible codelist?

In [31]:
sql = """
select count(distinct r.response_id)
from response_codelists r join codelist_statuses c on c.codelist = r.codelist
where c.status != 200;
"""
df = pd.read_sql(sql, engine)
df

count
0  16129

So 16,129 out of 19,689 ISO records have at least one reference to a downed codelist. 82%.

Question: Do you want to know what the downed codelists affect (role codes or progress codes or ??)? 

On to answering that question. Starting with which elements (the terminal element tag names).

In [6]:
# list of elements with number of responses containing at least reference for that element 
# pointing to one of the linkrot codelists (it is for a status code greater than 200, 
# knowing that there are no redirects in the set)

sql = """
with c as 
(
	select id, codelist
	from codelist_statuses
	where status > 200
), x as
(
	select r.id as response_id, r.host, r.metadata_age,
		lower(replace(left(i.codelist, strpos(i.codelist, '#')), '#', '')) as codelist,
		i.xpath, string_to_array(i.xpath, '/') as xpath_arr
	from codelists i join responses r on r.source_url_sha = i.file
)
select count(distinct x.response_id) as num_responses, 
	-- x.host, 
	-- x.metadata_age::date, 
	-- c.codelist,
	-- x.xpath, 
	x.xpath_arr[array_upper(xpath_arr, 1)] as last_x
from x join c on c.codelist = x.codelist
--order by x.codelist
group by last_x
order by num_responses DESC;
"""
    
df = pd.read_sql(sql, engine)
df

num_responses                               last_x
0           15079                          CI_RoleCode
1           12001                         MD_ScopeCode
2            9950                      CI_DateTypeCode
3            8229          MD_MaintenanceFrequencyCode
4            7061                   MD_KeywordTypeCode
5            6657                CI_OnLineFunctionCode
6            4828                  MD_CharacterSetCode
7            3792               DS_AssociationTypeCode
8            3614                   MD_RestrictionCode
9            3145                      MD_ProgressCode
10           1946                         LanguageCode
11           1769                      SV_CouplingType
12           1738                DS_InitiativeTypeCode
13           1606             MD_DimensionNameTypeCode
14           1268           MD_CoverageContentTypeCode
15           1068                  MD_CellGeometryCode
16            369              CI_PresentationFormCode
17            241     MD_SpatialRepresentationTypeCode
18            139                MD_ClassificationCode
19             78           MD_GeometricObjectTypeCode
20             28                    MD_MediumNameCode
21             25                         MX_ScopeCode
22             13                              DCPList
23             12                  MD_MediumFormatCode
24              8                              Country
25              7                CI_onLineFunctionCode
26              5                 MD_TopologyLevelCode
27              5                 MI_OperationTypeCode
28              3  GC_Security_Classification_TypeCode
29              3                             fileType
30              3                  EC_Program_TypeCode
31              3              EC_Directorate_TypeCode
32              3                   EC_Branch_TypeCode
33              3                      SDN_CountryCode
34              3               SDN_DeviceCategoryCode
35              3           SDN_HierarchyLevelNameCode
36              3           SDN_ParameterDiscoveryCode
37              3             SDN_PlatformCategoryCode
38              2                   SDN_FormatNameCode
39              2                          SDN_CRSCode
40              1                        SDN_EDMEDCode
41              1                         SDN_PortCode
42              1                      SDN_MarsdenCode
43              1                    SDN_WaterBodyCode
44              1                 MD_TopicCategoryCode
45              1          DQ_EvaluationMethodTypeCode
46              1                          SDN_CSRCode
47              1                 SDN_DataCategoryCode
48              1                      SDN_CSRUnitCode
49              1                     SDN_PlatformCode

In [3]:
# and by the full unqualified xpath
sql = """
with c as 
(
	select id, codelist
	from codelist_statuses
	where status > 200
), x as
(
	select r.id as response_id, r.host, r.metadata_age,
		lower(replace(left(i.codelist, strpos(i.codelist, '#')), '#', '')) as codelist,
		i.xpath, string_to_array(i.xpath, '/') as xpath_arr
	from codelists i join responses r on r.source_url_sha = i.file
)
select count(distinct x.response_id) as num_responses, 
	x.xpath
from x join c on c.codelist = x.codelist
group by x.xpath
order by num_responses DESC;
"""
    
df = pd.read_sql(sql, engine)
    
with pd.option_context('display.max_colwidth', 1000, 'display.max_rows', 310):
    display(df[:309])

num_responses  \
0             6569   
1             6437   
2             6369   
3             5814   
4             5732   
5             5725   
6             5607   
7             5581   
8             5374   
9             5210   
10            4778   
11            4261   
12            3849   
13            3213   
14            3149   
15            3127   
16            3053   
17            2995   
18            2984   
19            2774   
20            2632   
21            2624   
22            2481   
23            2466   
24            2360   
25            1879   
26            1850   
27            1794   
28            1746   
29            1726   
30            1722   
31            1720   
32            1702   
33            1649   
34            1631   
35            1620   
36            1594   
37            1587   
38            1537   
39            1345   
40            1336   
41            1330   
42            1314   
43            1156   
44            1141   
45            1134   
46            1054   
47             989   
48             859   
49             709   
50             705   
51             671   
52             651   
53             410   
54             395   
55             342   
56             284   
57             281   
58             272   
59             266   
60             253   
61             251   
62             217   
63             197   
64             173   
65             166   
66             160   
67             114   
68             114   
69             114   
70             114   
71             111   
72             108   
73             107   
74             104   
75             104   
76             104   
77             101   
78             101   
79             101   
80             101   
81             100   
82              98   
83              94   
84              94   
85              94   
86              94   
87              93   
88              91   
89              91   
90              91   
91              91   
92              91   
93              91   
94              86   
95              74   
96              70   
97              70   
98              69   
99              69   
100             68   
101             64   
102             62   
103             62   
104             59   
105             54   
106             49   
107             48   
108             47   
109             46   
110             42   
111             39   
112             38   
113             36   
114             34   
115             30   
116             29   
117             28   
118             26   
119             25   
120             24   
121             24   
122             24   
123             23   
124             22   
125             21   
126             19   
127             18   
128             18   
129             17   
130             17   
131             17   
132             17   
133             16   
134             16   
135             16   
136             16   
137             15   
138             15   
139             15   
140             14   
141             14   
142             14   
143             13   
144             13   
145             13   
146             12   
147             10   
148             10   
149             10   
150             10   
151             10   
152             10   
153             10   
154             10   
155             10   
156             10   
157             10   
158             10   
159             10   
160             10   
161             10   
162              9   
163              9   
164              9   
165              9   
166              9   
167              9   
168              9   
169              8   
170              8   
171              8   
172              8   
173              8   
174              7   
175              7   
176              6   
177              5   
178              5   
179              5   
180              5   
1

In [5]:
# and the percentages (based on a total of 19,689)
df['pct'] = (df['num_responses'] / 19689.) * 100.

with pd.option_context('display.max_colwidth', 1000, 'display.max_rows', 310):
    display(df[:100])

num_responses  \
0            6569   
1            6437   
2            6369   
3            5814   
4            5732   
5            5725   
6            5607   
7            5581   
8            5374   
9            5210   
10           4778   
11           4261   
12           3849   
13           3213   
14           3149   
15           3127   
16           3053   
17           2995   
18           2984   
19           2774   
20           2632   
21           2624   
22           2481   
23           2466   
24           2360   
25           1879   
26           1850   
27           1794   
28           1746   
29           1726   
30           1722   
31           1720   
32           1702   
33           1649   
34           1631   
35           1620   
36           1594   
37           1587   
38           1537   
39           1345   
40           1336   
41           1330   
42           1314   
43           1156   
44           1141   
45           1134   
46           1054   
47            989   
48            859   
49            709   
50            705   
51            671   
52            651   
53            410   
54            395   
55            342   
56            284   
57            281   
58            272   
59            266   
60            253   
61            251   
62            217   
63            197   
64            173   
65            166   
66            160   
67            114   
68            114   
69            114   
70            114   
71            111   
72            108   
73            107   
74            104   
75            104   
76            104   
77            101   
78            101   
79            101   
80            101   
81            100   
82             98   
83             94   
84             94   
85             94   
86             94   
87             93   
88             91   
89             91   
90             91   
91             91   
92             91   
93             91   
94             86   
95             74   
96             70   
97             70   
98             69   
99             69   

                                                                                                                                                                                                                                                                xpath  \
0                                                                                                                                                                                                            MI_Metadata/contact/CI_ResponsibleParty/role/CI_RoleCode   
1                                                                                                                                MI_Metadata/identificationInfo/MD_DataIdentification/citation/CI_Citation/citedResponsibleParty/CI_ResponsibleParty/role/CI_RoleCode   
2                                                                                                                                     MI_Metadata/distributionInfo/MD_Distribution/distributor/MD_Distributor/distributorContact/CI_ResponsibleParty/role/CI_RoleCode   
3                                                                                                                                                                                                                             MD_Metadata/hierarchyLevel/MD_ScopeCode   
4                                                                                                                                                 MI_Metadata/metadataMaintenance/MD_MaintenanceInformation/maintenanceAndUpdateFrequency/MD_MaintenanceFrequencyCode   
5                                                                                                                                                                                        MI_Metadata/dataQualityInfo/DQ_DataQuality/scope/DQ_Scope/level/MD_ScopeCode   
6                          

### Distribution URL Characterization

What can we understand about the URL provided, especially for the distributions? 

<img src="https://geo-ide.noaa.gov/wiki/images/0/03/CI_OnlineResource.png" title="iso_online_resource"/>

Image. From the [geo-ide wiki](https://geo-ide.noaa.gov/wiki/index.php?title=File:CI_OnlineResource.png), the structure of the CI_OnlineResource. 


In [5]:
# for function codes in distribution elements
sql = """
select element_value, count(element_value) as num_elems
from iso_onlineresources
where element_key = 'function_code' and tags ilike '%%/distributionInfo%%'
group by element_key, element_value
order by element_key, element_value, num_elems DESC;
"""

df = pd.read_sql(sql, engine)
df

element_value  num_elems
0                                                 375        113
1                                                 381         69
2                                            browsing        171
3                                             display         10
4                                            download      13352
5                                downloadRegistration          2
6                                          fileAccess          1
7   http://www.ngdc.noaa.gov/metadata/published/xs...        499
8                                         information       3527
9                                          mapDigital       1666
10                                      offlineAccess         30
11                                              order        198
12                                             search        585
13                                                URL          2
14                                               view          1
15                                         webService        110

In [6]:
# for protocols in distribution elements
sql = """
select element_value, count(element_value) as num_elems
from iso_onlineresources
where element_key = 'protocol' and tags ilike '%%/distributionInfo%%'
group by element_key, element_value
order by element_key, element_value, num_elems DESC;
"""

df = pd.read_sql(sql, engine)
df

element_value  num_elems
0                                           clip         12
1                                            DAP         20
2                                     directRead         53
3                                   EEA:FILEPATH          2
4                                 ERDDAP:griddap         96
5                                ERDDAP:tabledap         48
6                                           ESRI         28
7                     ESRI:MAP-1.0-rest-get-open          2
8                                           file          2
9                                            ftp        127
10                                           FTP        351
11                                          http       1959
12                                          HTTP       1272
13                                      HTTP/1.1         37
14                                 HTTP-DOWNLOAD          2
15                                         https        141
16                                           KML          1
17                                      NOAA:LAS         79
18                                       OGC:WCS          1
19                                       OGC:WFS         31
20                OGC:WFS-1.0.0-http-get-feature          2
21           OGC:WFS-2.0.0-http-get-capabilities         14
22                    OGC:WFS-2.0.0-http-get-map          2
23                                       OGC:wms          8
24                                       OGC:WMS         22
25                    OGC:WMS-1.1.0-http-get-map          2
26                    OGC:WMS-1.1.1-http-get-map         80
27           OGC:WMS-1.3.0-http-get-capabilities         17
28                    OGC:WMS-1.3.0-http-get-map         52
29          OGC:WMTS-1.0.0-http-get-capabilities          4
30                                       OPENDaP          1
31                                         order       1854
32                                          read         96
33                                       THREDDS         21
34  urn:ogc:serviceType:WebMapService:1.1.1:HTTP          2
35                WWW:DOWNLOAD-1.0-ftp--download          1
36               WWW:DOWNLOAD-1.0-http--download        196
37                                      WWW:LINK        369
38                       WWW:LINK-1.0-http--link       1707
39             WWW:LINK-1.0-http--link--download         60
40                    WWW:LINK-1.0-http--related        327

In [8]:
# for applicationProfiles in distribution elements
sql = """
select element_value, count(element_value) as num_elems
from iso_onlineresources
where element_key = 'application_profile' and tags ilike '%%/distributionInfo%%'
group by element_key, element_value
order by element_key, element_value, num_elems DESC;
"""

df = pd.read_sql(sql, engine)
df

element_value  num_elems
0                                       Analysis Tool          5
1                                      Any FTP client        248
2                                             browser        158
3                                             CSV/TXT         84
4                             Data Search Application         87
5                                       Documentation         55
6                                            Download        120
7                                        Google Earth        392
8                                                http          7
9   http://stategeothermaldata.org/uri-gin/aasg/xm...          2
10  http://stategeothermaldata.org/uri-gin/aasg/xm...          2
11  http://stategeothermaldata.org/uri-gin/aasg/xm...          1
12                                        Information         38
13                                                KML         77
14                             Map Search Application         74
15                                              Other          3
16                                               REST         50
17                                             Sample         88
18                                            Service          1
19                          Standard Internet browser        617
20                          Standard Internet Browser         68
21  Standard Internet browsers can browse OPeNDAP ...         18
22  Standard Internet browsers can browse THREDDS ...         18
23                                            THREDDS         33
24                                            Various          1
25                  VENDOR=ADOBE : APPLICATION=Reader         10
26               VENDOR = ADOBE: APPLICATION = Reader          2
27                   VENDOR=ESRI : APPLICATION=ArcGIS          3
28               VENDOR = ESRI: APPLICATION = ArcInfo          1
29               VENDOR = ESRI: APPLICATION = ARCINFO         48
30                      VENDOR=ESRI : APPLICATION=OGS          6
31                     VENDOR=ESRI : APPLICATION=OGS;          1
32  VENDOR=ESRI : APPLICATION=OGS; VENDOR=ADOBE : ...          2
33           VENDOR=MICROSOFT : APPLICATION=Word 2011          2
34                                        web browser        595
35                                        Web Browser        611
36                                                WMS          2
37                                                XLS          2
38                                                ZIP          6
39                                      Zipped Access          2

Not many unique strings in any of those three areas. And nothing about use in ISO records, ie limited to a few systems, etc.

In [4]:
# protocol terms count by host
sql = """
with j as 
(
	select response_id, element_value, count(id) as num_values
	from iso_onlineresources
	where element_key = 'protocol'
		and tags ilike '%%/distributionInfo%%'
	group by element_key, response_id, element_value
)
select element_value, 
	--round(avg(j.num_values), 2) as avg_values,
	r.host,
	count(distinct j.response_id) as num_responses
from j join responses r on j.response_id = r.id
group by r.host, element_value
order by element_value, r.host, num_responses DESC;
"""
df = pd.read_sql(sql, engine)
with pd.option_context('display.max_colwidth', 1000, 'display.max_rows', 150):
    display(df)

element_value  \
0                                            clip   
1                                             DAP   
2                                             DAP   
3                                             DAP   
4                                      directRead   
5                                      directRead   
6                                    EEA:FILEPATH   
7                                  ERDDAP:griddap   
8                                 ERDDAP:tabledap   
9                                            ESRI   
10                                           ESRI   
11                     ESRI:MAP-1.0-rest-get-open   
12                                           file   
13                                            ftp   
14                                            ftp   
15                                            ftp   
16                                            ftp   
17                                            ftp   
18                                            FTP   
19                                            FTP   
20                                            FTP   
21                                            FTP   
22                                            FTP   
23                                            FTP   
24                                            FTP   
25                                           http   
26                                           http   
27                                           http   
28                                           http   
29                                           http   
30                                           http   
31                                           http   
32                                           http   
33                                           http   
34                                           http   
35                                           http   
36                                           http   
37                                           http   
38                                           http   
39                                           http   
40                                           http   
41                                           HTTP   
42                                           HTTP   
43                                           HTTP   
44                                           HTTP   
45                                           HTTP   
46                                           HTTP   
47                                           HTTP   
48                                       HTTP/1.1   
49                                  HTTP-DOWNLOAD   
50                                          https   
51                                          https   
52                                            KML   
53                                       NOAA:LAS   
54                                        OGC:WCS   
55                                        OGC:WFS   
56                                        OGC:WFS   
57                                        OGC:WFS   
58                 OGC:WFS-1.0.0-http-get-feature   
59            OGC:WFS-2.0.0-http-get-capabilities   
60                     OGC:WFS-2.0.0-http-get-map   
61                                        OGC:wms   
62                                        OGC:WMS   
63                                        OGC:WMS   
64                                        OGC:WMS   
65                     OGC:WMS-1.1.0-http-get-map   
66                     OGC:WMS-1.1.1-http-get-map   
67                     OGC:WMS-1.1.1-http-get-map   
68                     OGC:WMS-1.1.1-http-get-map   
69                     OGC:WMS-1.1.1-http-get-map   
70                     OGC:WMS-1.1.1-http-get-map   
71            OGC:WMS-1.3.0-http-get-capabilities   
72            OGC:WMS-1.3.0-http-get-capabilities   
73            OGC:WMS-1.3.0-http-get-capabilities   
74                     OGC:WMS-1.3.0-http-get-map   
75      

In [5]:
# and for function codes

sql = """
with j as 
(
	select response_id, element_value, count(id) as num_values
	from iso_onlineresources
	where element_key = 'function_code'
		and tags ilike '%%/distributionInfo%%'
	group by element_key, response_id, element_value
)
select element_value,
	r.host,
	count(distinct j.response_id) as num_responses
from j join responses r on j.response_id = r.id
group by r.host, element_value
order by element_value, r.host, num_responses DESC;
"""
df = pd.read_sql(sql, engine)
with pd.option_context('display.max_colwidth', 1000, 'display.max_rows', 150):
    display(df)

element_value  \
0                                                                                                                 375   
1                                                                                                                 375   
2                                                                                                                 381   
3                                                                                                                 381   
4                                                                                                            browsing   
5                                                                                                            browsing   
6                                                                                                             display   
7                                                                                                             display   
8                                                                                                             display   
9                                                                                                            download   
10                                                                                                           download   
11                                                                                                           download   
12                                                                                                           download   
13                                                                                                           download   
14                                                                                                           download   
15                                                                                                           download   
16                                                                                                           download   
17                                                                                                           download   
18                                                                                                           download   
19                                                                                                           download   
20                                                                                                           download   
21                                                                                                           download   
22                                                                                                           download   
23                                                                                                           download   
24                                                                                                           download   
25                                                                                                           download   
26                                                                                                           download   
27                                                                                                           download   
28                                                                                                           download   
29                                                                                                           download   
30                                                                                                           download   
31                                                                                                           download   
32                                                                                                           do

So here, for rows 60-64, those are bugs in the ISO (the codeList and codeListValue attributes are flipped).